src/protego.py

```diff
@@ -0,0 +1,386 @@
+import re
+from datetime import time
+from collections import namedtuple
+import six
+from six.moves.urllib.parse import urlparse, urlunparse, ParseResult, quote, unquote
+import logging
+
+logger = logging.getLogger(__name__)
+
+Rule = namedtuple('Rule', ['field', 'value'])
+RequestRate = namedtuple(
+    'RequestRate', ['requests', 'seconds', 'start_time', 'end_time'])
+
+DISALLOW_DIRECTIVE = {'disallow', 'dissallow', 'dissalow', 'disalow', 'diasllow', 'disallaw'}
+ALLOW_DIRECTIVE = {'allow'}
+USER_AGENT_DIRECTIVE = {'user-agent', 'useragent', 'user agent'}
+SITEMAP_DIRECTIVE = {'sitemap', 'sitemaps', 'site-map'}
+CRAWL_DELAY_DIRECTIVE = {'crawl-delay', 'crawl delay'}
+REQUEST_RATE_DIRECTIVE = {'request-rate', 'request rate'}
+HOST_DIRECTIVE = {'host'}
+
+WILDCARDS = {'*', '$'}
+
+__all__ = ['RequestRate', 'Protego']
+
+
+class _URLPattern(object):
+    """Internal class which represents a URL pattern."""
+
+    def __init__(self, pattern):
+        self._pattern = pattern
+        self.priority = len(pattern)
+        self._contains_asterisk = '*' in self._pattern
+        self._contains_dollar = self._pattern.endswith('$')
+
+        if self._contains_asterisk:
+            self._pattern_before_asterisk = self._pattern[:self._pattern.find('*')]
+        elif self._contains_dollar:
+            self._pattern_before_dollar = self._pattern[:-1]
+
+        self._pattern_compiled = False
+
+    def match(self, url):
+        """Retun True if pattern matches the given URL, otherwise return False."""
+        # check if pattern is already compiled
+        if self._pattern_compiled:
+            return self._pattern.match(url)
+
+        if not self._contains_asterisk:
+            if not self._contains_dollar:
+                # answer directly for patterns without wildcards
+                return url.startswith(self._pattern)
+
+            # pattern only contains $ wildcard.
+            return url == self._pattern_before_dollar
+
+        if not url.startswith(self._pattern_before_asterisk):
+            return False
+
+        self._pattern = self._prepare_pattern_for_regex(self._pattern)
+        self._pattern = re.compile(self._pattern)
+        self._pattern_compiled = True
+        return self._pattern.match(url)
+
+    def _prepare_pattern_for_regex(self, pattern):
+        """Return equivalent regex pattern for the given URL pattern."""
+        pattern = re.sub(r'\*+', '*', pattern)
+        s = re.split(r'([$*])', pattern)
+        for index, substr in enumerate(s):
+            if substr not in WILDCARDS:
+                s[index] = re.escape(substr)
+            else:
+                s[index] = s[index].replace('*', '.*')
```

Comment:
Also, since the original robots.txt specification did not support using `$` as a marker of the end of the string, maybe we should allow for dollar signs to have both meanings?

For example, `Allow: /$` allows both just `/` and anything starting with `/$` (that is, `/%24`).